File: Week_4_Functions/visualized.py
import customtkinter as ctk
from matplotlib.figure import Figure
from matplotlib.backends.backend_tkagg import FigureCanvasTkAgg
from PIL import Image


class HappinessCalculator(ctk.CTk):
    def __init__(self):
        super().__init__()

        self.title("Happiness Level Calculator")
        self.geometry("1000x700")  # Increased width to accommodate the image

        self.grid_columnconfigure(0, weight=1)
        self.grid_columnconfigure(1, weight=1)
        self.grid_rowconfigure(0, weight=1)

        self.left_frame = ctk.CTkFrame(self)
        self.left_frame.grid(row=0, column=0, padx=10, pady=10, sticky="nsew")

        self.right_frame = ctk.CTkFrame(self)
        self.right_frame.grid(row=0, column=1, padx=10, pady=10, sticky="nsew")

        self.left_frame.grid_columnconfigure(0, weight=1)
        self.left_frame.grid_columnconfigure(1, weight=1)

        self.hours_relaxing = ctk.DoubleVar(value=4)
        self.hours_socializing = ctk.DoubleVar(value=2)
        self.hours_working = ctk.DoubleVar(value=8)

        self.create_sliders()
        self.create_results_display()
        self.create_chart()
        self.create_message_display()
        self.create_image_display()

        self.update_results()

    def create_sliders(self):
        sliders = [
            ("Hours Relaxing", self.hours_relaxing),
            ("Hours Socializing", self.hours_socializing),
            ("Hours Working", self.hours_working)
        ]

        for i, (label, variable) in enumerate(sliders):
            ctk.CTkLabel(self.left_frame, text=label).grid(row=i, column=0, padx=10, pady=5, sticky="w")
            slider = ctk.CTkSlider(self.left_frame, from_=0, to=24, number_of_steps=48, variable=variable)
            slider.grid(row=i, column=1, padx=10, pady=5, sticky="ew")
            slider.bind("<ButtonRelease-1>", self.update_results)

    def create_results_display(self):
        self.result_label = ctk.CTkLabel(self.left_frame, text="Happiness Level: 0", font=("Arial", 16, "bold"))
        self.result_label.grid(row=3, column=0, columnspan=2, padx=10, pady=10)

    def create_chart(self):
        self.figure = Figure(figsize=(6, 4), dpi=100)
        self.ax = self.figure.add_subplot(111)
        self.canvas = FigureCanvasTkAgg(self.figure, master=self.left_frame)
        self.canvas.get_tk_widget().grid(row=4, column=0, columnspan=2, padx=10, pady=10, sticky="nsew")

    def create_message_display(self):
        self.message_label = ctk.CTkLabel(self.left_frame, text="", font=("Arial", 22), wraplength=500)
        self.message_label.grid(row=5, column=0, columnspan=2, padx=10, pady=10)

    def create_image_display(self):
        # Replace 'path_to_your_image.png' with the actual path to your image file
        image_path = '/Users/hamiltonn/PycharmProjects/ConceptualCodingAssignment/Week_4_Functions/HappinessGoal.png'

        try:
            image = Image.open(image_path)
            photo = ctk.CTkImage(light_image=image, size=(400, 400))  # Adjust size as needed
            image_label = ctk.CTkLabel(self.right_frame, image=photo, text="")
            image_label.pack(expand=True, fill="both")
        except FileNotFoundError:
            error_label = ctk.CTkLabel(self.right_frame, text="Image not found. Please check the file path.",
                                       wraplength=200, font=("Arial", 12))
            error_label.pack(expand=True)

    def update_results(self, event=None):
        relax_score = self.hours_relaxing.get() * 2
        social_score = self.hours_socializing.get() * 3
        work_score = self.hours_working.get() * -2
        work_life_balance_score = work_score + relax_score
        happiness_level = (relax_score + social_score + work_life_balance_score) // 3

        self.result_label.configure(text=f"Happiness Level: {happiness_level:.0f}")

        self.update_chart(relax_score, social_score, work_life_balance_score)
        self.update_message(happiness_level)

    def update_chart(self, relax_score, social_score, work_life_balance_score):
        self.ax.clear()
        scores = [relax_score, social_score, work_life_balance_score]
        labels = ['Relax Score\n(Hours Relaxing * 2)',
                  'Social Score\n(Hours Socializing * 3)',
                  'Work Life Balance\n(Work Score + Relax Score)']
        bars = self.ax.bar(labels, scores)
        self.ax.set_ylabel('Score')
        self.ax.set_title('Component Scores')

        for bar in bars:
            height = bar.get_height()
            self.ax.text(bar.get_x() + bar.get_width() / 2, height,
                         f'{height:.1f}', ha='center', va='bottom')

        self.figure.tight_layout()
        self.canvas.draw()

    def update_message(self, happiness_level):
        message = f"Your Happiness Rating is: {happiness_level:.0f}.\n"
        if happiness_level >= 10:
            message += "You are very happy! Keep up the good work!"
        else:
            message += "You could be happier. Try to improve your scores in each category."
        self.message_label.configure(text=message)


if __name__ == "__main__":
    app = HappinessCalculator()
    app.mainloop()

File: Week_4_Functions/Visualized.js
import React, { useState, useEffect } from 'react';
import { Slider } from '@/components/ui/slider';
import { Card, CardContent, CardHeader, CardTitle } from '@/components/ui/card';
import { BarChart, Bar, XAxis, YAxis, Tooltip, ResponsiveContainer } from 'recharts';

const HappinessCalculator = () => {
  const [hoursRelaxing, setHoursRelaxing] = useState(4);
  const [hoursSocializing, setHoursSocializing] = useState(2);
  const [hoursWorking, setHoursWorking] = useState(8);
  const [happinessLevel, setHappinessLevel] = useState(0);

  const calculateHappiness = () => {
    const relaxScore = hoursRelaxing * 2;
    const socialScore = hoursSocializing / 2;
    const workScore = hoursWorking * -2;
    const workLifeBalanceScore = workScore + relaxScore;
    const happiness = Math.floor((relaxScore + socialScore + workLifeBalanceScore) / 3);
    setHappinessLevel(happiness);
  };

  useEffect(() => {
    calculateHappiness();
  }, [hoursRelaxing, hoursSocializing, hoursWorking]);

  const data = [
    { name: 'Relax Score', value: hoursRelaxing * 2 },
    { name: 'Social Score', value: hoursSocializing / 2 },
    { name: 'Work Life Balance', value: (hoursWorking * -2) + (hoursRelaxing * 2) },
  ];

  return (
    <div className="p-4 max-w-3xl mx-auto">
      <h1 className="text-2xl font-bold mb-4">Happiness Level Calculator</h1>
      <div className="grid grid-cols-1 md:grid-cols-2 gap-4">
        <Card>
          <CardHeader>
            <CardTitle>Input Values</CardTitle>
          </CardHeader>
          <CardContent>
            <div className="space-y-4">
              <div>
                <label className="block mb-2">Hours Relaxing: {hoursRelaxing}</label>
                <Slider
                  value={[hoursRelaxing]}
                  onValueChange={(value) => setHoursRelaxing(value[0])}
                  max={24}
                  step={0.5}
                />
              </div>
              <div>
                <label className="block mb-2">Hours Socializing: {hoursSocializing}</label>
                <Slider
                  value={[hoursSocializing]}
                  onValueChange={(value) => setHoursSocializing(value[0])}
                  max={24}
                  step={0.5}
                />
              </div>
              <div>
                <label className="block mb-2">Hours Working: {hoursWorking}</label>
                <Slider
                  value={[hoursWorking]}
                  onValueChange={(value) => setHoursWorking(value[0])}
                  max={24}
                  step={0.5}
                />
              </div>
            </div>
          </CardContent>
        </Card>
        <Card>
          <CardHeader>
            <CardTitle>Results</CardTitle>
          </CardHeader>
          <CardContent>
            <div className="text-center mb-4">
              <h2 className="text-xl font-semibold">Happiness Level</h2>
              <p className="text-4xl font-bold">{happinessLevel}</p>
            </div>
            <ResponsiveContainer width="100%" height={200}>
              <BarChart data={data}>
                <XAxis dataKey="name" />
                <YAxis />
                <Tooltip />
                <Bar dataKey="value" fill="#8884d8" />
              </BarChart>
            </ResponsiveContainer>
          </CardContent>
        </Card>
      </div>
    </div>
  );
};

export default HappinessCalculator;